### DISCUSSION TITLE
2033

### DISCUSSION THREAD

**Discussion by KC5VDJ** [kej715/DtCyber#153] [General]
These reserved bits look tempting.  A couple of words and one bit should be all that is needed.  The bit indicating "new time" vs "old time".  Bit not set? Use old code for time.  I'm thinking of a modular solution.  Iirc there is at least one bit available in the fs catalog as well.

I think this is worthy of the use of such reserved but never used words and bits..  I'm not referring to "installation-reserved", such a fundamental needed change deserves to go into what cdc reserved for future stuff in nos in the first place, not stomping on installations using the appropriate spots.

**Comment by kej715** [comment]
"Reserved" doesn't necessarily mean "unused" or "initialized to zero", so
it would be prudent to do some thorough testing before using a reserved
field, to ensure that it isn't actually used. And, if it can be verified
that it's not used and it's not initialized to zero, it would obviously be
necessary to introduce code that does so.

**Reply by kej715** [reply]
Or, perhaps the issue is that packed date values returned by macros such as PDATE contain year values that are offset from 1970? In that case, though, wouldn't the next problematic year by 2070, not 2033?

**Comment by KC5VDJ** [comment]
Yeah, trying to figure out what is truly unused is the "fun" part...

**Comment by wschaub** [comment]
I've given a lot of thought to 2033. no matter what is done it's going to be a lot of work because of all the various utilities/products and their file formats and how they store dates there.  The OS itself would probably be the more easy part as long as you provided the new style date information via a new OS request (and allowed the originals to stop at 2033 to prevent anything breaking that doesn't know about the new call)   

There's quite a lot of code to go through, and things like PLATO also have their own unique issues.  That said I would really like to see post 2033 support happen but it's non-trivial, particularly because in a lot of cases the date takes up space in a packed word and you simply can't make it bigger without totally changing the data structure in an established file format.

**Reply by kej715** [reply]
Please refresh my memory about why 2033 is a problem for NOS 2?  NOS 2 does not keep time relative to anything like the Unix Epoch. The RTCL field in CMR maintains a time value that is relative to the last deadstart, and the RTIME macro and associated system call returns that value. CMR does not have any fields that, for example, keep time relative to a specific date in history. The RTCL value can certainly rollover, but only when the system has not been re-deadstarted in more than 2**36 msecs. What am I overlooking?

**Reply by wschaub** [reply]
I believe the big problem is in the PDTL word in CMR   
![image](https://github.com/kej715/DtCyber/assets/955669/fe8983e5-6233-4a87-b975-3f02fe35408a)

NOS only allocates 6 bits biased by 1970 for the year and at some point in 2033 it runs out. If you look at the modsets on the y2k.tap tape you will see commentary that addresses this issue. the NOS maintainers solution in the modsets was to make sure that time stops advancing when we hit our maximum and display a DSD message.  

now PDTL has some bits that are reserved to 0 that we might be able to then use, but the catch to that is auditing everything that 
returns data from PDTL and uses that data directly in PP code because it might expect those bits to be zero.

Also there's data structures for the filesystem that expect to use that packed date/time information with zero room to make that 6 bit year larger.  I believe we have similar issues in things like program library formats, etc. 

I hope this helps. 

We also track the julian date which I haven't looked into any issues there but as far as I know when you ask for the real date PDTL is the word that is used, unless you just want it in display code which is also in separate words in low core.

**Comment by KC5VDJ** [comment]
Exactly.  Thus the old/new flag idea. If set, use the new fields.  If not set (assumed default), don't.  Effectively, 48 bits gives three million years off the top of the head, thus one 60 bit word, stored as integer seconds since 19700101000000.0. 

No matter how it happens, it's a guaranteed pita.  I think it's worth the effort.  I also think waiting until 1999, er, 2032 isn't that bright.  Better to get it out of the way early, lesson of 1999.

**Comment by kej715** [comment]
Duh! That makes sense now. I had miscalculated the length of the packed
year returned by PDATE (which simply returns PDTL).

**Comment by KC5VDJ** [comment]
Outside of secured systems, how common was it to not clear out "garbage data" (used by something else, freed, yet not cleared) from memory?

I'm largely referring to unused fields on storage media not being set before being written, thus containing said "garbage data".

If this was a thing, compatibility would require more than one bit for a flag, more like a multi-bit unlikely pattern to distinguish itself from trash.

The assumption is the obvious one: the default is exactly as things are now, with newtime requiring either this flag or a unique pattern.  If unique pattern, obviously the more bits the better, maybe even a word.  If twelve bits can do the trick, only a single contiguous word is necessary for the whole thing, assuming a 48 bit second counter (just over 3 million years, which should be more than adequate).  Not as long as the modern 64 bit counters, but it'll be least cost (cycles).  Unless anyone has met Dr. Who, and has inside info, let 'em figure it out again three million years from now...

**Comment by kej715** [comment]
We probably don't need a new counter for time.  The NOS real time clock
does not reflect an offset from a specific historical date or time. The NOS
real time clock simply tracks the number of milliseconds (and seconds) that
have elapsed since the last deadstart. The current 36-bit value is probably
sufficient because it allows for a system to be running continuously for
more than 2 years before it will roll over.

The 2033 problem is due to the fact that packed dates have a 6-bit field
for the year, and they are currently biased by 1970.  1970 + 63 = 2033. So,
after 2033, the 6-bit field in packed dates will overflow.  Expanding the
subfield holding the packed year from 6 bits to 8 bits would allow for a
256 year range instead of a 64 year range. Expanding it to 12 bits would
allow for millennia. The PDTL field in CMR could allow for this expansion
since its upper 24 bits are explicitly defined as being 0. Of course, a
thorough review of the codebase would be needed to find and handle cases
where PP or CP programs expect those upper 24 bits to be zero and misbehave
if they're not.

Perhaps a more challenging problem will be dealing with other places where
packed dates are stored and expected to have 6-bit year values, e.g., files
and catalog entries. Finding extra bits in places like catalog entries and
the system sector of files might not be possible.

Another possibility might be to enable the 6-bit packed year to roll over,
back to 0 when 2034 arrives. Perhaps the display code value in DTEL could
be used in deciding how to convert a packed date to displayable form (i.e.,
whether to bias it by 1970 or 2024). Of course, this might cause stored
objects with very old dates, e.g., files created in the early 80's, to be
displayed with incorrect dates, but that might be acceptable.
